src/main/java/app/attestation/server/AttestationServer.java

```diff
@@ -1010,6 +1012,89 @@ public void handlePost(final HttpExchange exchange) throws IOException, SQLiteEx
         }
     }
 
+    private static class AttestationHistoryHandler extends PostHandler {
+        @Override
+        public void handlePost(final HttpExchange exchange) throws IOException, SQLiteException {
+            int page;
+            String fingerprint;
+            final Account account = verifySession(exchange, false, null);
+            if (account == null) {
+                return;
+            }
+            try (final JsonReader reader = Json.createReader(exchange.getRequestBody())) {
+                final JsonObject object = reader.readObject();
+                fingerprint = object.getString("fingerprint");
+                page = Integer.valueOf(object.getString("page"));
+                if (page < 0) {
+                    throw new NumberFormatException("Invalid page provided");
+                }
+                if (!verifyDeviceFingerprint(account.userId, fingerprint)) {
+                    throw new NullPointerException("Invalid device fingerprint provided");
+                }
+            } catch (final ClassCastException | JsonException | NullPointerException | NumberFormatException e) {
+                e.printStackTrace();
+                exchange.sendResponseHeaders(400, -1);
+                return;
+            }
+            writeAttestationHistoryJson(exchange, fingerprint, page);
+        }
+    }
+
+    private static boolean verifyDeviceFingerprint(final long userId, final String deviceFingerprint)
+            throws IOException, SQLiteException {
+        boolean isFingerprintAuthenticated = false;
+        final SQLiteConnection conn = new SQLiteConnection(AttestationProtocol.ATTESTATION_DATABASE);
+        try {
+            final SQLiteStatement select = conn.prepare("SELECT fingerprint FROM Devices WHERE userId is ?");
+            select.bind(1, userId);
+
+            while (select.step()) {
+                String ourFingerprint = BaseEncoding.base16().encode(select.columnBlob(0));
+                if (ourFingerprint.equals(deviceFingerprint)) {
+                    isFingerprintAuthenticated = true;
+                    break;
+                }
+            }
+
+            if (!isFingerprintAuthenticated) {
+                throw new NullPointerException("Device fingerprint does not belong to account");
+            }
+        } finally {
+            conn.dispose();
+        }
+        return isFingerprintAuthenticated;
+    }
+
+    private static void writeAttestationHistoryJson(final HttpExchange exchange, final String deviceFingerprint,
```

Comment:
Make sure that the indexes being used for pagination are stable. The oldest entry should be treated as 0 so that it never changes. The devices.json API could return the number of entries / pages so that the client knows how to fetch the newest first.